packages/neo-one-client-core/src/sc/createSmartContract.ts

```diff
@@ -69,53 +69,81 @@ const getParamsAndOptions = ({
   readonly hash?: Hash256String;
 } => {
   const hasRest = parameters.length > 0 && parameters[parameters.length - 1].rest;
-  const hasForwardOptions =
-    parameters.length > 0 &&
-    parameters[parameters.length - 1].rest &&
-    parameters[parameters.length - 1].type === 'ForwardValue';
-  let lastArgIndex = hasRest ? parameters.length - 1 : parameters.length;
-
-  let params = args;
-
-  let transfer: Transfer | undefined;
-  let hash: Hash256String | undefined;
-  if (send || completeSend || refundAssets) {
-    const maybeLastArg = params[lastArgIndex];
-    if (send) {
-      transfer = argAssertions.assertTransfer('transfer', maybeLastArg);
-    } else {
-      hash = argAssertions.assertHash256('hash', maybeLastArg);
-    }
-    params = params.slice(0, lastArgIndex).concat(params.slice(lastArgIndex + 1));
-  }
 
-  const maybeForwardOptions = params[lastArgIndex];
-  // tslint:disable-next-line no-any
-  let forwardOptions: any = {};
-  if (common.isOptionsArg(maybeForwardOptions) && hasForwardOptions) {
-    params = params.slice(0, lastArgIndex).concat(params.slice(lastArgIndex + 1));
-    forwardOptions = maybeForwardOptions;
-    lastArgIndex -= 1;
-  }
+  const { requiredArgs, fValues, forwardOptionsIn, optionsIn, transfer, hash } = args.reduceRight<{
+    // tslint:disable-next-line:no-any
+    readonly requiredArgs: ReadonlyArray<any>;
+    // tslint:disable-next-line:no-any
+    readonly fValues: ReadonlyArray<any>;
+    // tslint:disable-next-line:no-any
+    readonly optionsIn: any;
+    readonly forwardOptionsIn: ForwardOptions | undefined;
+    readonly transfer: Transfer | undefined;
+    readonly hash: Hash256String | undefined;
+  }>(
+    (acc, right) => {
+      // this block is just for rest param ones, potentially others I think.
+      if (common.isForwardValue(right) && acc.requiredArgs.length <= 0 && hasRest) {
+        return {
+          ...acc,
+          fValues: [right].concat(acc.fValues),
+        };
+      }
 
-  const maybeOptionsArg = params[lastArgIndex] as {} | undefined;
-  // tslint:disable-next-line no-any
-  let optionsIn: any = {};
-  if (common.isOptionsArg(maybeOptionsArg)) {
-    params = params.slice(0, lastArgIndex).concat(params.slice(lastArgIndex + 1));
-    // tslint:disable-next-line no-any
-    optionsIn = maybeOptionsArg as any;
-    lastArgIndex -= 1;
-  }
+      if (common.isForwardValueOptions(right) && acc.forwardOptionsIn === undefined) {
+        return {
+          ...acc,
+          forwardOptions: right,
```

Comment:
Hmmmmm, something doesn't seem right here....

Maybe don't name any of the accumulators with the `In` suffix??